src/arrow1/reader_async.rs

```diff
@@ -1,15 +1,670 @@
+        use wasm_bindgen_futures::JsFuture;
+        let (sender, receiver) = oneshot::channel();
+        let file = self.inner.clone();
+        spawn_local(async move {
+            let subset_blob = file
+                .slice_with_i32_and_i32(
+                    range.start.try_into().unwrap(),
+                    range.end.try_into().unwrap(),
+                )
+                .unwrap();
+            let buf = JsFuture::from(subset_blob.array_buffer()).await.unwrap();
+            let out_vec = Uint8Array::new_with_byte_offset(&buf, 0).to_vec();
+            sender.send(out_vec).unwrap();
+        });
+
+        receiver.await.unwrap()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct JsFileReader {
+    file: WrappedFile,
+    coalesce_byte_size: usize,
+}
+
+impl JsFileReader {
+    pub fn new(file: web_sys::File, coalesce_byte_size: usize) -> Self {
+        Self {
+            file: WrappedFile::new(file.into()),
+            coalesce_byte_size,
+        }
+    }
+}
+
+impl AsyncFileReader for JsFileReader {
+    fn get_bytes(&mut self, range: Range<usize>) -> BoxFuture<'_, parquet::errors::Result<Bytes>> {
+        async {
+            let (sender, receiver) = oneshot::channel();
+            let mut file = self.file.clone();
+            spawn_local(async move {
+                let result: Bytes = file.get_bytes(range).await.into();
+                sender.send(result).unwrap()
+            });
+            let data = receiver.await.unwrap();
+            Ok(data)
+        }
+        .boxed()
+    }
+
+    fn get_byte_ranges(
+        &mut self,
+        ranges: Vec<Range<usize>>,
+    ) -> BoxFuture<'_, parquet::errors::Result<Vec<Bytes>>> {
+        let fetch_ranges = merge_ranges(&ranges, self.coalesce_byte_size);
+
+        // NOTE: This still does _sequential_ requests, but it should be _fewer_ requests if they
+        // can be merged.
+        async move {
+            let mut fetched = Vec::with_capacity(ranges.len());
+
+            for range in fetch_ranges.iter() {
+                let data = self.get_bytes(range.clone()).await?;
+                fetched.push(data);
+            }
+
+            Ok(ranges
+                .iter()
+                .map(|range| {
+                    // a given range CAN span two coalesced row group sets.
+                    // log!("Range: {:?} Actual length: {:?}", range.end - range.start, res.len());
+                    let idx = fetch_ranges.partition_point(|v| v.start <= range.start) - 1;
+                    let fetch_range = &fetch_ranges[idx];
+                    let fetch_bytes = &fetched[idx];
+
+                    let start = range.start - fetch_range.start;
+                    let end = range.end - fetch_range.start;
+                    fetch_bytes.slice(start..end)
+                })
+                .collect())
+        }
+        .boxed()
+    }
+
+    fn get_metadata(&mut self) -> BoxFuture<'_, parquet::errors::Result<Arc<ParquetMetaData>>> {
```

Comment:
It would be nice if we could deduplicate this with the HTTP metadata fetching a bit.